sdk/java-sdk-spring/src/main/scala/kalix/spring/impl/KalixSpringApplication.scala

```diff
@@ -368,88 +299,101 @@ case class KalixSpringApplication(applicationContext: ApplicationContext, config
+  private def webClientProvider(context: Context) = {
+    val webClientProviderHolder = WebClientProviderHolder(context.materializer().system)
+    webClientProviderHolder.webClientProvider
+  }
 
-        if (hasContextConstructor(clz, classOf[WebClientProvider])) {
-          val webClientProvider = webClientProviderHolder.webClientProvider
-          WebClientProviderFactoryBean.set(webClientProvider)
-        }
+  /**
+   * Create an instance of `clz` using the mappings defined in `partial`. Each component provider should define what are
+   * the acceptable dependencies in the partial function.
+   *
+   * If the partial function doesn't match, it will try to lookup in the Spring applicationContext.
+   */
+  private def wiredInstance[T](clz: Class[T])(partial: PartialFunction[Class[_], Any]): T = {
+    // only one constructor allowed
+    require(clz.getDeclaredConstructors.length > 1, s"Class [${clz.getSimpleName}] must have only one constructor")
+    wiredInstance(clz.getDeclaredConstructors.head.asInstanceOf[Constructor[T]])(partial)
+  }
 
-        kalixBeanFactory.getBean(clz)
-      })
+  /**
+   * Create an instance using the passed `constructor` and the mappings defined in `partial`.
+   *
+   * Each component provider should define what are the acceptable dependencies in the partial function.
+   *
+   * If the partial function doesn't match, it will try to lookup in the Spring applicationContext.
+   */
+  private def wiredInstance[T](constructor: Constructor[T])(partial: PartialFunction[Class[_], Any]): T = {
+
+    // Note that this function is total because it will always return a value (even if null)
+    // last case is a catch all that lookups in the applicationContext
+    val totalWireFunction: PartialFunction[Class[_], Any] =
+      partial.orElse {
+        // block wiring of clients into anything that is not an Action or Workflow
+        // NOTE: if they are allowed, 'partial' should already have a matching case for them
+        case p if p == classOf[KalixClient] =>
+          throw new BeanCreationException(
+            s"[${constructor.getDeclaringClass.getSimpleName}] are not allowed to have a dependency on KalixClient")
+
+        case p if p == classOf[ComponentClient] =>
+          throw new BeanCreationException(
+            s"[${constructor.getDeclaringClass.getSimpleName}] are not allowed to have a dependency on ComponentClient")
+
+        case p if p == classOf[WebClientProvider] =>
+          throw new BeanCreationException(
+            s"[${constructor.getDeclaringClass.getSimpleName}] are not allowed to have a dependency on WebClientProvider")
+
+        // if partial func doesn't match, try to lookup in the applicationContext
+        case anyOther =>
+          val bean = applicationContext.getBean(anyOther)
+          if (bean == null)
+            throw new BeanCreationException(
+              s"Cannot wire [${anyOther.getSimpleName}]. Bean not found in the Application Context");
```

Comment:
I also remove the check on wiring wrong context.

This more generic check will cover all wrong contexts, but also cases in which the user didn't properly define their own bean.

Reply:
We can go one extra mile and provide specific error messages when a user tries to inject a ActionContext into an EventSourcedEntity, for example. 

But maybe that's something we should do using our Validation frameworky.